crates/bevy_ecs/src/resource.rs

```diff
@@ -72,4 +86,123 @@ pub use bevy_ecs_macros::Resource;
     label = "invalid `Resource`",
     note = "consider annotating `{Self}` with `#[derive(Resource)]`"
 )]
-pub trait Resource: Send + Sync + 'static {}
+pub trait Resource: Component<Mutability = Mutable> {
+    /// The `on_add` component hook that maintains the uniqueness property of a resource.
+    fn on_add_hook(mut deferred_world: DeferredWorld, context: HookContext) {
```

Comment:
I remember that #21346 had a way to re-use these hooks for dynamic resources, but in this version it looks like they're only available on the `Resource` trait.  Do we want to make these free functions and try to plug them in when registering a dynamic resource?  

For that matter, even if we don't plug them in automatically for dynamic resources, these don't seem to depend on the `Self` type and can be free functions instead of associated functions on the trait.

Reply:
I followed the example of the `Relationship` trait, which keeps the hooks under the trait. I don't see the use of them being free functions, as I can plug them into the dynamic resource as-is through `Resource::on_add_hook`, I just like to organize them on the trait they're related to.

Reply:
> I don't see the use of them being free functions

One reason is to make them more convenient to use for dynamic resources.  

Another is to avoid monomorphization.  Putting them on the trait means they get compiled separately for every `Resource` type, while a non-generic free function would only be compiled once.  

In contrast, the `Relationship` trait uses things like `Self::RelationshipTarget`, so it really does need to be compiled separately for each relationship.